src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+    std::set<Bond> ring_bonds;
+    
+    while (!all(hit_atoms)) {
+        auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
+        auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
+        auto& current_atom_bonds = adj_list[current_atom];
+        hit_atoms[current_atom] = true;
+        if (current_atom_bonds.empty()) {
+            continue;
+        }
+
+        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms,
+                          current_atom_bonds[0], current_atom
+        );
+    }
+
+    return ring_atoms;
+}
+
+static void write_atom_smiles(std::unique_ptr<TextFile>& file_, const Atom& atom) {
+    bool needs_brackets = false;
+    auto type = atom.type();
+
+    // The mass must be an integer is the only check we need as all atoms in the
+    // periodic table do have non-integer masses. Therefore, if the the mass is
+    // not an integer, then we know the user has set an isotope
+    double mass;
+    size_t mass_int = 0;
+    if (std::modf(atom.mass(), &mass) == 0.0 && mass != 0.0) {
+        needs_brackets = true;
+        mass_int = static_cast<size_t>(mass);
+    }
+
+    // If any of these values are set / not a default value
+    auto smi_class = atom.get("smiles_class");
+    auto chirality = atom.get<Property::STRING>("chirality").value_or("");
+    auto explicit_h =atom.get<Property::DOUBLE>("hydrogen_count").value_or(0.0);
+
+    // Charge is similar to mass. It must be an integer, otherwise it is ignored
+    double charge;
+    int charge_int = 0;
+    if (std::modf(atom.charge(), &charge) == 0.0 && charge != 0.0) {
+        needs_brackets = true;
+        charge_int = static_cast<int>(charge);
+    }
+
+    // Before the atom is printed, we must know if we are printing a property set.
+    if (explicit_h != 0.0 || smi_class || !chirality.empty() ||
+        aliphatic_organic.count(type) == 0) {
+        needs_brackets = true;
+    }
+
+    auto is_aromatic = atom.get<Property::BOOL>("is_aromatic").value_or(false);
+
+    if (is_aromatic) {
+        // We need to print brackets for aromatic Te, Se, As, and Si
+        // Note, we allow aromatic Bromine to be bare (no brackets) in following
+        // ChemAxon's standard.
+        if (type.size() > 1) {
+            needs_brackets = true;
+        }
+        std::transform(type.begin(), type.end(), type.begin(), ::tolower);
+    }
+
+    if (needs_brackets) {
+        fmt::print(*file_, "[");
+    }
+
+    // Mass must be first, before the element is printed
+    if (mass_int != 0) {
+        fmt::print(*file_, "{}", mass_int);
+    }
+
+    fmt::print(*file_, "{}", type);
+
+    if (smi_class && smi_class->kind() == Property::DOUBLE) {
+        fmt::print(*file_, ":{}", static_cast<int>(smi_class->as_double()));
+    }
+
+    if (chirality != "") {
+        // If it's clockwise, then we just print the symbol @@ and be done with it
+        if (chirality.find("CW") == 0) {
```

Comment:
Is `--- CW foobar` a valid chirality flag? I think this code would accept it.

Reply:
It would accept `CW foobar` (the check is to see if the string begins with `CW`). I put a note in the documentation that we do not check chirality - at all. So if the user sets something funny, they'll get a funny result. I can add warnings, however.

Reply:
> I can add warnings, however.

This would be nice, to remove possible sources of confusion :smiley: